Remove app pause function and state

jost-s opened the pull request:
### Summary

resolves #5092 

### TODO:
- [x] CHANGELOGs updated with appropriate info
- [x] All code changes are reflected in docs, including module-level docs

veeso commented on crates/holochain_types/src/app.rs:
```diff
@@ -406,7 +406,7 @@ impl From<RunningApp> for InstalledApp {
 }
 
-/// An app which is either Paused or Disabled, i.e. not Running
+/// An app which is Disabled, i.e. not Running
```
If this is related to `AppStatus::Disabled`, to help users navigate the documentation, you could change to "[`AppStatus::Disabled`]".

veeso commented on crates/holochain/CHANGELOG.md:
```diff
@@ -9,6 +9,7 @@ The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/).
 
 - Test app operations (install/enable/disable/uninstall) with regards to app state and cell state.
 - **BREAKING CHANGE**: Remove `start_app` from conductor. Use `enable_app` instead.
+- **BREAKING CHANGE**: Remove `pause_app` and `Paused` state from conductor. Pausing an app was used when enabling an app partially failed and could be re-attempted. Now the app is only enabled if all cells successfully started up.
```
If this is a breaking change, the commit message should not be a `chore` by convention, I think, but it should be either a feat or fix with the breaking flag `!`

matthme commented on crates/holochain/src/conductor/conductor.rs:
```diff
@@ -394,7 +394,7 @@ mod startup_shutdown_impls {
             // We don't care what fx are returned here, since all cells need to
             // be spun up
-            let _ = self.start_paused_apps().await?;
+            let _ = self.start_apps().await?;
```
I haven't looked into the code deeply yet but that doesn't seem correct to me here. I don't think we want to start stopped apps when initializing the conductor as these include apps that have been deliberately disabled by the user. Do we have a test that checks that an app can be disabled via the adminwebsocket and won't get automatically re-enabled on conductor restart?

jost-s replied:
I see what you mean. Yeah, this is no longer applicable. Apps can't be paused, so they won't be started when initializing the conductor.

matthme commented on crates/holochain/src/conductor/conductor.rs:
```diff
@@ -721,10 +721,10 @@ mod dna_impls {
         /// Restart every paused app
         #[cfg_attr(feature = "instrument", tracing::instrument(skip_all))]
-        pub(crate) async fn start_paused_apps(&self) -> ConductorResult<AppStatusFx> {
+        pub(crate) async fn start_apps(&self) -> ConductorResult<AppStatusFx> {
```
If this function is starting stopped apps I would prefer it would be called `start_stopped_apps()`, otherwise it sounds like it is starting all apps. Or is this what it's doing?

ThetaSinner commented on crates/holochain/src/conductor/conductor.rs:
```diff
@@ -721,10 +718,10 @@ mod dna_impls {
         /// Restart every paused app
         #[cfg_attr(feature = "instrument", tracing::instrument(skip_all))]
-        pub(crate) async fn start_paused_apps(&self) -> ConductorResult<AppStatusFx> {
+        pub(crate) async fn start_stopped_apps(&self) -> ConductorResult<AppStatusFx> {
```
Has the last usage of this not been removed above?

Semantically, starting a paused app is quite different from pausing a stopped app. We aren't distinguishing between failed to start and deliberately stopped by the user. I think that's fine but then we shouldn't just start all stopped apps, we should let the user selectively start the apps they want?

I think this can be removed, if it can't, I think it might make sense to try to remove it.

ThetaSinner commented on crates/holochain_conductor_api/src/admin_interface.rs:
```diff
@@ -633,21 +633,12 @@ impl ExternalApiWireError {
-    Running,
-    /// Filter on apps which are Stopped, i.e. not Running.
-    /// This includes apps in the Disabled status, as well as the Paused status.
-    Stopped,
```
Ahh, I remember now. So stopped is a composite state and we should be removing references to "stopped" in favor of "disabled"

matthme replied:
Actually, I'm seeing that the [StoppedAppReason](https://github.com/holochain/holochain/blob/a7ca376ac98d709c0594b52952696fe22ef54742/crates/holochain_types/src/app.rs#L1097-L1105) contains the `AwaitingMemproofs` variant which I'm not sure we can conflate with `Disabled` completely. But I'm also not entirely sure how `StoppedAppReason` relates to `AppStatusFilter`.

jost-s replied:
@matthme Did you mean to say "should not be"?

ThetaSinner reviewed (commented):
Still finding the terminology a little confusing. If I'm following right the an installed app can either be enabled or disabled. An enabled app can either be running or stopped depending on whether its cells were able to start?

ThetaSinner commented on crates/holochain/src/conductor/conductor.rs:
```diff
@@ -2032,8 +2010,6 @@ mod app_status_impls {
         /// Adjust app statuses (via state transitions) to match the current
         /// reality of which Cells are present in the conductor.
         /// - Do not change state for Disabled apps. For all others:
-        /// - If an app is Paused but all of its (required) Cells are on,
-        ///   then set it to Running
         /// - If an app is Running but at least one of its (required) Cells are off,
         ///   then set it to Paused
```
Right, then this can be ignored.

jost-s commented:
@matthme @ThetaSinner Thanks both of you for the hints. I've done another sweep and removed `stopped` and `started` as statuses and replaced with `disabled` and `enabled` where applicable in a separate commit.

It was well possible to include in this PR and not postpone until the bigger refactor.

matthme reviewed (approved):
Looks good to me know :+1:
